### PR TITLE
feat: Serialize duration as string, not as nanosecs, in API responses and DB

### PR DESCRIPTION
**Pull request by michaljurecko**
Part of: https://keboola.atlassian.net/browse/PSGO-388

**Changes:**
- **`time.Duration` is by default serialized as `int64`, as number of nanoseconds.**
- It is acceptable for database, but it complicates E2E tests.
- **It is not acceptable for API, we had some conversions implemented on the API layer.**
- **Now, we have in the API much more `time.Duration` values that need to be handled.**
- So I created `duration.Duration` type, that serializes as string, for example: `100ms`, `1h0m`, ...
- It is not implemented in the Go in this ways, to keep BC:
  - https://github.com/golang/go/issues/10275
  - > This isn't possible to change now, as it would change the encodings produced by programs that exist today.
- We have similar package for the time serialization, see [utctime](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-387-duration-pkg/internal/pkg/service/common/utctime/utctime.go) pkg.

-----------

### CODE REVIEW

File: internal/pkg/service/common/duration/duration.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Package duration provides a wrapper for time.Duration type,
+// to serialize duration as string instead of int64 nanoseconds.
+//
+// This cannot be changed in Go 1.X, it would break backward compatibility, see:
+// https://github.com/golang/go/issues/10275
+// "This isn't possible to change now, as it would change the encodings produced by programs that exist today."
+package duration
```
Comment (michaljurecko) [inline internal/pkg/service/common/duration/duration.go:7]: Docs.

File: internal/pkg/service/common/duration/duration.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func (v *Duration) UnmarshalJSON(b []byte) (err error) {
+	// String, for example, "1h20s"
+	if bytes.HasPrefix(b, []byte(`"`)) && bytes.HasSuffix(b, []byte(`"`)) {
+		return v.UnmarshalText(bytes.Trim(b, `"`))
+	}
+	// Nanoseconds int64 - backward compatibility
+	return json.Unmarshal(b, (*time.Duration)(v))
+}
```
Comment (michaljurecko) [inline internal/pkg/service/common/duration/duration.go:49]: Unmarshaling is backward compatible, nanoseconds as an int are valid.

File: internal/pkg/service/buffer/definition/repository/sink_repo_test.go [hunk 3]
```diff
@@ -470,10 +470,10 @@ definition/sink/active/123/456/my-source-1/my-sink-1
...
             "mode": "disk",
             "wait": false,
-            "checkInterval": 1000000,
+            "checkInterval": "1ms",
             "countTrigger": 100,
             "bytesTrigger": "100KB",
-            "intervalTrigger": 100000000
+            "intervalTrigger": "100ms"
           }
         }
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/repository/sink_repo_test.go:478]: Change in the database snapshot.

File: internal/pkg/service/buffer/sink/tablesink/statistics/collector/collector_test.go [hunk 4]
```diff
@@ -37,7 +37,7 @@ func TestCollector(t *testing.T) {
 	syncCounter := 0
 	triggerSyncAndWait := func() {
 		syncCounter++
-		clk.Add(cfg.SyncInterval)
+		clk.Add(cfg.SyncInterval.Duration())
 		assert.Eventually(t, func() bool {
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/sink/tablesink/statistics/collector/collector_test.go:41]: Updated in related code.

File: internal/pkg/service/buffer/config/config_test.go [hunk 5]
```diff
@@ -106,15 +106,15 @@ sink:
                 diskSync:
                     # Sync mode: "disabled", "cache" or "disk". Validation rules: required,oneof=disabled disk cache
                     mode: disk
-                    # Wait for sync to disk OS cache or to disk hardware, depending on the mode. Validation rules: excluded_if= Mode disabled
+                    # Wait for sync to disk OS cache or to disk hardware, depending on the mode.
                     wait: true
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/config/config_test.go:110]: Unrelated: Fixed in the config test, the rules have been updated in a prev PR, tests don't run in the CI yet.

**Review by hosekpeter — APPROVED** [review]
LGTM

File: internal/pkg/service/buffer/sink/tablesink/statistics/config.go [hunk 6]
```diff
@@ -16,22 +18,22 @@ type Config struct {
 }
 
 type SyncConfig struct {
-	SyncInterval time.Duration `configKey:"interval" configUsage:"Statistics synchronization interval, from memory to the etcd." validate:"required,minDuration=100ms,maxDuration=5s"`
-	SyncTimeout  time.Duration `configKey:"timeout" configUsage:"Statistics synchronization timeout."  validate:"required,minDuration=1s,maxDuration=1m"`
+	SyncInterval duration.Duration `configKey:"interval" configUsage:"Statistics synchronization interval, from memory to the etcd." validate:"required,minDuration=100ms,maxDuration=5s"`
+	SyncTimeout  duration.Duration `configKey:"timeout" configUsage:"Statistics synchronization timeout."  validate:"required,minDuration=1s,maxDuration=1m"`
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/buffer/sink/tablesink/statistics/config.go:22]: No, if the struct is not used in db/api, `time.Duration` is ok.

File: internal/pkg/validator/custom.go [hunk 7]
```diff
@@ -139,17 +140,20 @@ func (v *wrapper) registerCustomRules() {
 		Rule{
 			Tag: "minDuration",
 			FuncCtx: func(ctx context.Context, fl validator.FieldLevel) bool {
-				value, ok := fl.Field().Interface().(time.Duration)
-				if !ok {
+				var value time.Duration
+				if v, ok := fl.Field().Interface().(time.Duration); ok {
+					value = v
+				} else if v, ok := fl.Field().Interface().(duration.Duration); ok {
+					value = v.Duration()
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/validator/custom.go:147]: Notes from call:
```
var value any = duration.Duration(123) // 123ns
var value any = int64(123)

--------------

if v, ok := value.(duration.Duration); ok {
  // ...
}

--------------

swtich v := value.(type) {
  case duration.Duration:
  case int64:
  default:
    panic(....)
}

```

**Pull request merged**